torch/csrc/autograd/python_function.cpp

```diff
@@ -849,35 +849,35 @@ PyObject *THPFunction_apply(PyObject *cls, PyObject *inputs)
 // Backward
 ////////////////////////////////////////////////////////////////////////////////
 
-static void _prepare_grad_output(THPFunction *self, THPObjectPtr& raw_grad_output)
+static void _prepare_grads(THPFunction *self, THPObjectPtr& raw_grads, bool is_grad_output)
 {
   AutoGPU gpu_guard(-1);
-  int num_grad_output = PyTuple_GET_SIZE(raw_grad_output.get());
-  // First, check if any of grad_outputs is None. If not, there's nothing to do
+  int num_grads = PyTuple_GET_SIZE(raw_grads.get());
+  // First, check if any of grads is None. If not, there's nothing to do
   bool has_none = false;
-  for (int i = 0; i < num_grad_output; i++) {
-    has_none |= PyTuple_GET_ITEM(raw_grad_output.get(), i) == Py_None;
+  for (int i = 0; i < num_grads; i++) {
+    has_none |= PyTuple_GET_ITEM(raw_grads.get(), i) == Py_None;
   }
   if (!has_none)
       return;
 
-  THPObjectPtr grad_output;
-  grad_output = PyTuple_New(num_grad_output);
-  if (!grad_output) throw python_error();
+  THPObjectPtr grads;
+  grads = PyTuple_New(num_grads);
+  if (!grads) throw python_error();
 
   // Look for Nones and replace them with new buffers
-  auto& output_info = self->output_info;
-  for (int i = 0; i < num_grad_output; i++) {
-    PyObject *grad = PyTuple_GET_ITEM(raw_grad_output.get(), i);
+  auto& grads_info = is_grad_output ? self->output_info : self->input_info;
```

Comment:
Can you add an assertion that `TORCH_ASSERT(grads_info.size() == (size_t)num_grads)`